src/scala/com/github/johnynek/bazel_deps/TryMerge.scala

```diff
@@ -1,51 +1,72 @@
 package com.github.johnynek.bazel_deps
 
+import cats.{Foldable, Reducible}
+import cats.data.NonEmptyList
 import scala.util.{Failure, Success, Try}
-import cats.implicits._
 
-trait TryMerge[T] {
+import cats.syntax.all._
+
+trait TryMerge[T] { self =>
   def tryMerge(debugName: Option[String], left: T, right: T): Try[T]
+  def reverse: TryMerge[T] =
+    new TryMerge[T] {
+      def tryMerge(debugName: Option[String], left: T, right: T): Try[T] = self.tryMerge(debugName, right, left)
+      override def reverse = self
+    }
 }
 
 object TryMerge {
   def tryMerge[T: TryMerge](debugName: Option[String], a: T, b: T): Try[T] =
     implicitly[TryMerge[T]].tryMerge(debugName, a, b)
 
+  def tryMergeAll[F[_]: Reducible, A: TryMerge](debugName: Option[String], r: F[A]): Try[A] =
+    r.reduceLeftM(Success(_): Try[A])(tryMerge(debugName, _, _))
+
+  def tryMergeAllOr[F[_]: Foldable, A: TryMerge](debugName: Option[String], r: F[A], res: => Try[A]): Try[A] =
+    NonEmptyList.fromList(r.toList) match {
+      case Some(nel) => tryMergeAll[NonEmptyList, A](debugName, nel)
+      case None => res
+    }
+
   implicit def tryOptMerge[T: TryMerge]: TryMerge[Option[T]] =
     new TryMerge[Option[T]] {
       def tryMerge(debugName: Option[String], left: Option[T], right: Option[T]): Try[Option[T]] = {
         (left, right) match {
           case (None, None)       => Success(None)
           case (Some(l), Some(r)) => TryMerge.tryMerge(debugName, l, r).map(Some(_))
-          case (Some(l), None)    => Success(Some(l))
-          case (None, Some(r))    => Success(Some(r))
+          case (l @ Some(_), None)    => Success(l)
+          case (None, r @ Some(_))    => Success(r)
         }
       }
     }
 
   implicit def tryStringMapMerge[T: TryMerge]: TryMerge[Map[String, T]] =
     new TryMerge[Map[String, T]] {
+      private val rev = reverse
       def tryMerge(
         debugName: Option[String],
           left: Map[String, T],
           right: Map[String, T]
       ): Try[Map[String, T]] = {
-        (left.keySet | right.keySet).toList.sorted.foldM(Map.empty[String, T]) {
-          (m, nextK) =>
-            val r: Try[T] =
-              (left.get(nextK), right.get(nextK)) match {
-                case (Some(l), None)    => Success(l)
-                case (None, Some(r))    => Success(r)
-                case (Some(l), Some(r)) =>
-                  TryMerge.tryMerge(Some(debugName.map{p => s"$p:$nextK"}.getOrElse(nextK)), l, r)
-                case (None, None) =>
-                    Failure(new Exception(s"Shouldn't happen, key=$nextK was in keyset"))
+        if (right.size > left.size) rev.tryMerge(debugName, right, left)
+        else {
+          // right <= left in size
+          val overlaps = right.exists { case (k, _) => left.contains(k) }
+          if (!overlaps) Success(left ++ right)
+          else {
+            right.toList.sortBy(_._1)
```

Comment:
Same comment as earlier. Also, merging `SortedMaps` to preserve the deterministic behavior might make the overall flow a bit easier.

Reply:
yeah, I think we should do that as a future change.